Use ClusterState as Consistency Source for Snapshot Repositories

original-brownbear opened the pull request:
Follow up to #49729 

This change removes falling back to listing out the repository contents to find the latest `index-N` in write-mounted blob store repositories.
This saves 2-3 list operations on each snapshot create and delete operation. Also it makes all the snapshot status APIs cheaper (and faster) by saving one list operation there as well in many cases.
This removes the resiliency to concurrent modifications of the repository as a result and puts a repository in a `corrupted` state in case loading `RepositoryData` failed from the assumed generation.

Note to reviewers:

I suggest starting from the new test ``CorruptedBlobStoreRepositoryIT` illustrating the mechanics of a repository being marked as "corrupted" to make sure we're in agreement on the behavior change here. I know we discussed in the snapshot meeting a while back, but let's make sure we're all in agreement on the tradeoffs made here:

* (good) Slightly faster and cheaper (cloud repos) repository operations (both dimensions are somewhat irrelevant I'd say for most users)

* (good-neutral) Concurrent modifications to the repository by multiple clusters are detected much more reliably and force the user to re-add the repo by hand, making it so that the dangerous situation of having two clusters write to the repo at the same time is impossible. This seems like mostly a win, but he off situation where a user may use the repo for writes across two clusters and manually ensures that operations don't run at the same time would be broken by this change

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

original-brownbear commented:
Jenkins run elasticsearch-ci/1

original-brownbear commented:
Need to get https://github.com/elastic/elasticsearch/pull/49514 in first to stabilize `SnapshotResiliencyTests` enough so that they don't become unstable from this change

original-brownbear commented:
Jenkins run elasticsearch-ci/1

original-brownbear commented:
Jenkins run elasticsearch-ci/2

original-brownbear committed:
Consistent RepositoryData Load on Blobstores

This moves the blob store repository to only use the information
available in the clusterstate for loading `RepositoryData` without
falling back to listing to determine a repositories' generation.

Relates #49729
Closes #38941

original-brownbear committed:
last step

original-brownbear committed:
fix tests

original-brownbear committed:
cs

original-brownbear committed:
fix tests

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into repo-uses-cs

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into repo-uses-cs

original-brownbear committed:
updated docs

original-brownbear committed:
simpler

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into repo-uses-cs

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -279,29 +294,44 @@ protected void doClose() {
         metadata = getRepoMetaData(state);
-        final long finalBestGen = Math.max(bestGenerationFromCS, metadata.generation());
-        latestKnownRepoGen.updateAndGet(known -> Math.max(known, finalBestGen));
+        if (bestEffortConsistency) {
```
This block can go away in `master` after backport to `7.x`, I'd keep the "old"-way of guessing from the in-progress actions for `7.x` though.

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -985,36 +1020,116 @@ public void endVerification(String seed) {
     // Tracks the latest known repository generation in a best-effort way to detect inconsistent listing of root level index-N blobs
     // and concurrent modifications.
-    private final AtomicLong latestKnownRepoGen = new AtomicLong(RepositoryData.EMPTY_REPO_GEN);
+    private final AtomicLong latestKnownRepoGen = new AtomicLong(RepositoryData.UNKNOWN_REPO_GEN);
```
Can't just start from the empty generation any more because we're not per-default falling back to listing.

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -985,36 +1020,116 @@ public void endVerification(String seed) {
+            if (clusterService.state().nodes().isLocalNodeElectedMaster() == false) {
+                // Non-master node does not load RepositoryData before master has at least loaded it once and thus initialized in the
+                // cluster state
+                listener.onFailure(new RepositoryException(metadata.name(), "Repository not initialized yet"));
```
Non-master nodes only ever load `RepositoryData`  in `org.elasticsearch.index.shard.StoreRecovery#restore` (in which case master will have loaded `RepositoryData` before to verify that the snapshot to be restored actually exists in the repo) as far as I can see so in the real world we should never get here (except for some freak concurrency cases where master fails over concurrently).

original-brownbear replied:
This actually becomes a non-issue thanks to your suggestion in https://github.com/elastic/elasticsearch/pull/49060#discussion_r357232276 since we can simply fall back to listing on the non-master nodes so long as the master hasn't initialized the latest generation in the CS

original-brownbear committed:
simpler

original-brownbear commented on test/framework/src/main/java/org/elasticsearch/repositories/AbstractThirdPartyRepositoryTestCase.java:
```diff
@@ -76,6 +76,7 @@ public void setUp() throws Exception {
     @Override
     public void tearDown() throws Exception {
         deleteAndAssertEmpty(getRepository().basePath());
+        client().admin().cluster().prepareDeleteRepository("test-repo").get();
```
Can't reuse the repo across tests that clear out its content any longer

original-brownbear committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into repo-uses-cs

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -985,36 +1020,116 @@ public void endVerification(String seed) {
+                listener.onFailure(new RepositoryException(metadata.name(), "Repository not initialized yet"));
+            } else {
+                // After initializing the repo in the cluster state we retry loading the latest RepositoryData on the generic pool
+                initializeRepoInClusterState(ActionListener.wrap(v ->
```
I don't think we should initialize this here (as it can be based on an inconsistent read). Instead I prefer we add this after a successful write repo operation and load leniently until that has happened.

original-brownbear replied:
Thanks, that's a sweet idea I think. That way we force at least one more listing before writing the new repo generation and are even safer in a freaky scenario where the repo was just modified and then mounted for writing :) Will adjust accordingly.

original-brownbear replied:
Removed the initialization logic in the read path completely now and simply made `BlobStoreRepository` use the old style listing fallback as long as the cluster state contains `RepositoryData#UNKNOWN_REPO_GEN` in the `RepositoryMetaData`. That also solves all the BwC issues IMO, since if we fall over from an older master, the cluster state will container `RepositoryData#UNKNOWN_REPO_GEN` and if we're already on a new version master then the contents of the cluster state will be correct if that master has ever written a new `index-N` or will be `RepositoryData#UNKNOWN_REPO_GEN` otherwise.

ywelsch reviewed (changes requested):
Thanks for tackling this, Armin. I've left some comments.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into repo-uses-cs

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -279,29 +294,44 @@ protected void doClose() {
     @Override
     public void updateState(ClusterState state) {
-        if (readOnly) {
+        bestEffortConsistency = isReadOnly() || state.nodes().getMinNodeVersion().before(RepositoryMetaData.REPO_GEN_IN_CS_VERSION);
```
You're right the upgrade experience is potentially broken here. I'll fix this together with the adjustments to the initialization.

> Also, I wonder if we should (at least in 7.x) allow a feature flag to disable enforcing consistency. At least to ensure we are not breaking existing use cases (e.g. something fishy on Cloud)

That sounds like a good idea. I wonder how to implement this though: should we do a system property? (then things could get a little weird if it's not set on all nodes but it seems like the easiest approach)

ywelsch replied:
I would prefer a deprecated cluster setting

original-brownbear replied:
Cluster setting it is -> will do :)

original-brownbear replied:
See my comment here https://github.com/elastic/elasticsearch/pull/49060#discussion_r357605538  on the BwC situation, that should be resolved now.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into repo-uses-cs

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -279,29 +294,44 @@ protected void doClose() {
+                // cluster state to a cluster that does track the generation in the cluster state
+                return;
+            }
+            final long previousBest = latestKnownRepoGen.getAndSet(metadata.generation());
```
My thinking was, that if we get here the cluster state is messed up in some form and we can't really change that anyway (it would be a bug in some cluster state update before) and I simply added this to have the debug logging + assertion. If we don't update the generation pointer on a node while the CS was incorrectly updated in some way we don't get anything out of that do we?

original-brownbear committed:
remove repository initialization

original-brownbear committed:
remove repository initialization

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -985,36 +1020,116 @@ public void endVerification(String seed) {
-        ActionListener.completeWith(listener, () -> {
-            // Retry loading RepositoryData in a loop in case we run into concurrent modifications of the repository.
-            while (true) {
+        if (latestKnownRepoGen.get() == RepositoryData.CORRUPTED_REPO_GEN) {
```
This is a non-issue now I think since with the removal of the initialization logic we're only really using the value once before we enter the read loop that spins on this value and actively expects and deals with it changing.

original-brownbear committed:
add setting overrid

original-brownbear committed:
fix test

original-brownbear commented:
@ywelsch thanks for taking a look. I addressed all your comments now I hope. I removed the repo initialization from the read path and added a deprecated setting to allow concurrent repo modifications like we currently have (see the new test added to `CorruptedBlobStoreRepositoryIT`).
I think this is good for another review :)

ywelsch commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -184,6 +185,14 @@
+     * Note: This setting is intended as a backwards compatibility solution for 7.x and will go away in 8.
+     */
+    public static final Setting<Boolean> ALLOW_CONCURRENT_MODIFICATION =
+        Setting.boolSetting("allow_concurrent_mod", false, Setting.Property.Deprecated, Setting.Property.Dynamic);
```
Let's fully spell out the name, i.e. `allow_concurrent_modifications`

ywelsch replied:
Also, do we need to define NodeScope but not dynamic, like we have for COMPRESS_SETTING?

original-brownbear replied:
> Also, do we need to define NodeScope but not dynamic, like we have for COMPRESS_SETTING?

Not really, all these attributes are just descriptive only as far as I can tell since these are not registered settings but and simply applied to the repository metadata. I'll remove the `Dynamic` part as it's pointless :)

ywelsch reviewed (commented):
I wonder if we should always be lenient w.r.t. to loading repository data when `pendingGeneration != safeGeneration`, in case where the pendingGeneration might have actually been successfully written to the repo (and possibly even index.latest been updated). It's possible that other read-only repos will observe that new state of the repo, and the writer to the repo will then bring it backwards (in particular risky if cluster state updates are slow at that particular point in time, and there is a longer lag between writing to the repo and updating safe repository).

original-brownbear commented:
>  It's possible that other read-only repos will observe that new state of the repo, and the writer to the repo will then bring it backwards (in particular risky if cluster state updates are slow at that particular point in time, and there is a longer lag between writing to the repo and updating safe repository).

Bringing the repo backwards is not really a possibiilty I think because if the cluster state update after the write failed it will be retried by the next master that will just finalize the still ongoing operation. That's exactly what was fixed in https://github.com/elastic/elasticsearch/pull/49995 I'd say. There we had tests failing because the repo was "taken back" because the snapshot job was removed from the CS without retrying the write.

If we add leniency here, it complicates things quite a bit IMO because we get situations like:

0. We have some hanging master that is stuck before writing index-N and just failed
1. Master A now tries to read non-existent index-N and destroys read-after-write consistency for that key then fails over for some reason
2. The hanging master comples the write to index-N
3. Whatever new master is elected may or may not see the just written `index-N` at the pending generation

(this would also fail existing tests for eventually consistent blob stores that explicitly make sure we never read from an inconsistent key)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into repo-uses-cs

original-brownbear committed:
rename an remove pointless prop

ywelsch commented:
> it will be retried by the next master that will just finalize the still ongoing operation

probably only unless you have a full cluster restart. In that case, SnapshotInProgress and friends are lost whereas RepositoriesMetaData is not.

I want to mainly understand the worst case here, and how that could possibly affect a user. In theory we could fix the concurrent master issue by appending the master term as an additional tie breaker to the file name.

original-brownbear commented:
> I want to mainly understand the worst case here, and how that could possibly affect a user. In theory we could fix the concurrent master issue by appending the master term as an additional tie breaker to the file name.

I think the worst case is as you mention the full cluster restart right before a snapshot finishes (i.e. before `RepositoryMetadata` generations get aligned). In that case, if you were to restore last snapshot taken and concurrently ran either a repository cleanup or snapshot delete the restore may fail mid-way because the cleanup or delete will interpret the last snapshot's files as unreferenced and remove them. 
I feel like this is a really fringe case though. Right now you have a bunch of failure scenarios in various ES versions when restoring while deleting from a repository (the step of reading `RepositoryData` which happens on the data nodes restoring is error prone in any version if a new one is written concurrently ... ) so I'm not sure it's worth it worrying about this? It's super unlikely and even without this change deleting anything while restoring has the potential to kill the restore, we're just adding one new very unlikely scenario.

original-brownbear commented:
> I want to mainly understand the worst case here, and how that could possibly affect a user. In theory we could fix the concurrent master issue by appending the master term as an additional tie breaker to the file name.

I think the worst case is as you mention the full cluster restart right before a snapshot finishes (i.e. before `RepositoryMetadata` generations get aligned). In that case, if you were to restore last snapshot taken and concurrently ran either a repository cleanup or snapshot delete the restore may fail mid-way because the cleanup or delete will interpret the last snapshot's files as unreferenced and remove them. 
I feel like this is a really fringe case though. Right now you have a bunch of failure scenarios in various ES versions when restoring while deleting from a repository (the step of reading `RepositoryData` which happens on the data nodes restoring is error prone in any version if a new one is written concurrently ... ) so I'm not sure it's worth it worrying about this? It's super unlikely and even without this change deleting anything while restoring has the potential to kill the restore, we're just adding one new very unlikely scenario.

> appending the master term as an additional tie breaker to the file name

So basically we'd have `index-{N}-{term}` and if we run into say pending generation `5`, we'd list `index-5-*` and move to that as the safe generation if we were to find a blob for `index-5-{someTerm}`? I think we could do that, but since it really only helps the concurrent repository access case I'm not sure it's worth the changes (we'd have to change what we store in `index.latest` and refactor the generation to be two `long` or a `String` wouldn't we?).

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into repo-uses-cs

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into repo-uses-cs

original-brownbear committed:
correct handling unclean full cluster restart

original-brownbear committed:
docs

original-brownbear commented:
@ywelsch as discussed earlier today, I added logic to identify a "dirty" startup of the repository now in e3f7ff2 (code and test) and 564202b (docs) that shoul make us never lose the latest state (assuming the list operation works well enough on full cluster restart :) but it should be fine and even if it isn't once in a blue moon, the repo can not become corrupted as a result still).

original-brownbear committed:
stable

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into repo-uses-cs

ywelsch reviewed (approved):
LGTM. Can you, in a follow-up, also add more tests for read-only repositories, given that they have a slightly different way to load stuff?

original-brownbear commented:
Thanks Yannick!

> Can you, in a follow-up, also add more tests for read-only repositories, given that they have a slightly different way to load stuff?

Will do :)
